### ISSUE TITLE
Problems loading bundled .mdl model files

### ISSUE THREAD

**Issue by tellypresence** [assimp/assimp#2898]
Have two different errors loading bundled .mdl model files:

Model file
`test/models/MDL/MDL (HL1)/man01.mdl`
fails to load with error
`File is too small.`

Model file
`test/models/MDL/MDL (HL1)/manT.mdl`
fails to load with error
`std::exception` (that's the entire error message)

**Comment by turol** [comment]
The code doesn'd handle exceptions correctly, after some tweaking I got a better error message:
```
[Half-Life 1 MDL] Model has no bodyparts
```
Looks like these models have always been broken. They were introduced by @malortie, any comments on whether they should be broken or not?

**Comment by malortie** [comment]
@tellypresence @turol 

Thanks for reporting. 

I am posting this to let everyone know I have been investigating this issue and found the root cause, which I explain in my next post. The error about the exception message will also be explained in the next post.

Starting with the 2 files you mentioned, **man01.mdl** and **manT.mdl**. The files are not broken. See below.

First, the entire model "man" is made of 3 files.

* man.mdl
* man01.mdl
* manT.mdl

**man.mdl** is the main MDL file. It contains meshes; bone info; sequence description eye position; model bounds. It also contains animation data for animations that are paired with the "default" sequence group, which is always the first sequence group.

**man01.mdl** is a demand loaded sequence group file. Demand loaded sequence group files store animation data such as bone position; rotation; compressed scale.

**manT.mdl** is an external MDL texture file. StudioMDL may store textures either in the main MDL file (man.mdl) or in an external MDL file, in this case (manT.mdl). A texture file contains texture pixels, texture info - width, height, palette.

For more information about the MDL file formats, please refer to this [page](https://github.com/malortie/assimp/wiki/MDL:-Half-Life-1-file-format).

----
Secondly, as for both files **man01.mdl** and **manT.mdl**.

**man01.mdl** is a demand loaded sequence group file and this file is split in small chunks containing the rotation and position for each frame of the animation, per bone. Please see the [MDL file format](https://github.com/malortie/assimp/wiki/MDL:-Half-Life-1-file-format#Animation-frame-offsets) for more information.

Unfortunately, a sequence group file does not store information about how many bones the model has, nor the offsets to each animation data chunk, so it's impossible to know the number of chunks, as well as the size of each chunk. Because of this, it is not possible to load an MDL from a demand loaded sequence group file. If you try to open man01.mdl using a Half-Life model viewer such as Jed's Half-Life model viewer, or from the original model viewer found in the Half-Life SDK, it will not work. Reading the original HLMV source code, it is assumed that the file loaded is the main MDL.

**manT.mdl** is a file used for storing textures externally. When writing an SMD file, the modeler may want to store the textures in an external file rather than in the main MDL itself by adding the command **$externaltextures**. In the case of **man.mdl**, I decided to store the textures in an external file to ensure it could load external MDL files. As of this moment, the Half-Life 1 MDL importer I wrote does not yet handle the case where a developer only wants to load textures from a Half-Life 1 model. Both Jed's Half-Life model viewer as well as the model viewer from the original Half-Life 1 SDK are capable of loading an external texture file (manT.mdl) without first loading the main MDL file (man.mdl).

In short, both files are not broken. They are just containers of additional data and they are read correctly as long as you import the main MDL file (man.mdl), which will then take care of loading the textures and the animations from external files. 

In the next post, I will explain my investigation about the issues you encountered.

**Comment by turol** [comment]
I adjusted exception handling as part of #2904 so it correctly passes the exception up the stack and to the user.

**Comment by malortie** [comment]
I am still writing the "next" post that will explain my investigations.

@tellypresence When loading the **manT.mdl**, you said you get an exception with the message `std::exception`. 

I was unable to reproduce this issue to get the message `std::exception`. 

When loading **manT.mdl**, the error message I always get in the exception is `[Half-Life 1 MDL] Model has no bodyparts` and correctly prints to output debugger when the code reaches line 140 in BaseImporter.cpp, when **ASSIMP_LOG_ERROR** is called.

@turol What are the tweaks you made to get the correct error message?

**Comment by turol** [comment]
https://github.com/assimp/assimp/pull/2904/commits/89e622060b02a5911bdabcb64afdd8a20c16c0e8

TL;DR: catch all exceptions and re-throw them without caring about their type.

**Comment by malortie** [comment]
In this post, I explain my investigation about the first issue posted by @tellypresence in which the MDL importer throws an exception `File is too small`.

Before I start explaining, here are a few definitions.

As explained in my first post, there are many types of MDL files for the model "man.mdl".

* **man.mdl** is the main MDL file.
* **man01.mdl** is a demand loaded sequence group file.
* **manT.mdl** is a file containing all textures for the MDL.

Also, the MDL header (Quake 1) has a size of 84 bytes. The size for a HL1 MDL sequence group header is 76 bytes.

### The cause of the issue

In **man.qc**, there is the following command: **$sequencegroupsize 64**. In QC files, it is possible to define new sequence groups but I do not define new groups, which means that the only sequence group I have is the "default" sequence group and is stored in the main MDL file.

Because of my instruction  **$sequencegroupsize 64**, this causes StudioMDL to change the maximum sequence group size ([See this file](https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/utils/studiomdl/studiomdl.c#L3356)) and later set the number of sequence groups to 2. [See this file](https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/utils/studiomdl/studiomdl.c#L1033-L1037). 

The algorithm runs and sets the sequence group index of each sequence to the "default" sequence group - the main MDL file. NOTE: The algorithm performs work for sequences that have activity 0, or ACT_RESET but it never does this because all my sequences use an activity other than 0. 

Finally, StudioMDL checks to see if it needs to create demand loaded sequence group files. Because the number of sequence groups is set to 2, it will create one sequence group file. [See this file](https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/utils/studiomdl/write.c#L537).  

The code writes the sequence group file header (76 bytes) and calls `WriteAnimations` but does not write any data because the sequence group index of each sequence has been replaced to 0. [See this file](https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/utils/studiomdl/write.c#L289-L291). 

Because each sequence have sequence group index set to 0, their animation data is saved in the main MDL file. [See this file](https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/utils/studiomdl/write.c#L624).

This causes the sequence group file **man01.mdl** to have a size of 76 bytes, which is less than the size of the Quake 1 file header. This issue occurs because of the way StudioMDL works, but the problem is not that the MDL file is broken or invalid (The Half-Life 1 game and the model viewers properly load **man.mdl** even if **man01.mdl** file with no animation data). 

The HL1 importer function in MDLLoader.cpp should throw an exception if the developer tries to load an MDL sequence group file directly but in this case, it's not possible because the file header size of a sequence group file is less than the Quake 1 file header size.

### Recommendation

The problem in this case is that the code in MDLLoader.cpp compares the size of the Quake 1 file header with the file size and the sequence group file header has a lesser size in bytes than the one from Quake 1.

There are 2 possible solutions:

1. Compare the file size against the sequence group file header size (**MDL::HalfLife::SequenceHeader_HL1**) instead of **MDL::Header**. 

2. Change the code so that these two could be independent. At the same time, this would allow the other MDL implementations to become independant.

Solution 1 is easier, but retains the problem of depending against the file size to consider the input file eligible to be processed further.

Solution 2 implies letting MDL implementations take care of reallocating the buffer themselves, after checking the file header magic word, themselves rather than having a centralized place as currently. Potentially increases the risks of having memory leaks.

Personally, I would prefer Solution 2 due to considerations of future MDL implementations - There are still unlisted MDL files such as Serious Sam's MDL models, but for safety I will choose Solution 1. I can take care of performing the required changes, but I would appreciate to have a bit of feedback on both solutions and see which one you would prefer.

**Comment by malortie** [comment]
> [89e6220](https://github.com/assimp/assimp/commit/89e622060b02a5911bdabcb64afdd8a20c16c0e8)
> 
> TL;DR: catch all exceptions and re-throw them without caring about their type.

@turol Thanks.

**Comment by malortie** [comment]
@tellypresence 

The first issue with `File is too small` has been solved. [See this PR](https://github.com/assimp/assimp/pull/2908).

**Comment by kimkulling** [comment]
Ok, I will close the issue.

**Comment by tellypresence** [comment]
Retried w/recent version of assimp:

Have two different errors loading bundled .mdl model files:

Model file
```
test/models/MDL/MDL (HL1)/man01.mdl
```
fails to load with error
`Impossible to properly load a model from an MDL sequence file`
(was `File is too small.`)

Error message much improved for
```
test/models/MDL/MDL (HL1)/manT.mdl
```
Message now:
`[Half-Life 1 MDL] Model has no bodyparts`

**Comment by malortie** [comment]
@tellypresence I made [some changes](https://github.com/assimp/assimp/pull/2952) to the HL1 MDL importer. It is now possible to load MDL texture files directly .e.g `manT.mdl`. Please note that only textures will be loaded.
